xds/src/main/java/com/linecorp/centraldogma/xds/internal/XdsResourceWatchingService.java

```diff
@@ -84,24 +88,33 @@ protected void init() {
             watchingGroups.add(groupName);
             final Revision normalizedRevision = repository.normalizeNow(Revision.HEAD);
             logger.info("Creating xDS resources from {} at revision: {}", groupName, normalizedRevision);
-            final Map<String, Entry<?>> entries = repository.find(normalizedRevision, pathPattern()).join();
-            for (Entry<?> entry : entries.values()) {
-                if (entry.type() != EntryType.JSON || !entry.hasContent()) {
-                    continue;
+            final CompletableFuture<Map<String, Entry<?>>> findFuture =
+                    repository.find(normalizedRevision, pathPattern());
+            futures.add(findFuture.handleAsync((entries, cause) -> {
+                // Executed by the repository worker thread.
+                if (cause != null) {
+                    logger.warn("Unexpected exception while finding {} at revision: {}",
+                                groupName, normalizedRevision, cause);
+                    return null;
                 }
-                final String path = entry.path();
-                final String contentAsText = entry.contentAsText();
-                try {
-                    handleXdsResource(path, contentAsText, groupName);
-                } catch (Throwable t) {
-                    throw new RuntimeException("Unexpected exception while building an xDS resource from " +
-                                               groupName + path, t);
+                for (Entry<?> entry : entries.values()) {
+                    if (entry.type() != EntryType.JSON || !entry.hasContent()) {
+                        continue;
+                    }
+                    final String path = entry.path();
+                    final String contentAsText = entry.contentAsText();
+                    try {
+                        handleXdsResource(path, contentAsText, groupName);
```

Comment:
In `handleXdsResource`, there is a heavy computation: serializing the text to a protobuf message.
Will revisit this later, if this PR doesn't make any changes.